runtime/component/dnd-mode.js

```diff
@@ -0,0 +1,634 @@
+    var s = property.get(END_TIME_KEY, 'persist')
+    if (s !== undefined) {
+      return s
+    } else {
+      return '8:00'
+    }
+  }
+
+  /**
+   * Set end time of the dnd mode
+   * @function getEndTime
+   * @param {string} endTime - the end time with hours and minutes (+8 tz for now)
+   */
+  static setEndTime (endTime) {
+    property.set(END_TIME_KEY, endTime, 'persist')
+  }
+
+  /**
+   * Set awake switch of the dnd mode
+   * @function getEndTime
+   * @param {string} awakeSwitch - awake switch 'open'/'close', set 'close' by default
+   */
+  static setAwakeSwitch (awakeSwitch) {
+    if (awakeSwitch !== 'open' && awakeSwitch !== 'close') {
+      awakeSwitch = 'close'
+    }
+    property.set(AWAKE_SWITCH_KEY, awakeSwitch, 'persist')
+  }
+
+  /**
+   * format start/end time 'hh:mm' to datetime today
+   * @function getEndTime
+   * @param {string} timeStr - time with hours and minutes eg '22:00'
+   * @param {number} defaultHour - default hours
+   * @param {number} defaultMinute - default minutes
+   * @returns {Date} the start/end time today
+   */
+  static formatTime (timeStr, defaultHour, defaultMinute) {
+    var d = new Date()
+    d.setSeconds(0)
+    d.setMilliseconds(0)
+    var hour = defaultHour
+    var minute = defaultMinute
+    try {
+      var array
+      if (typeof timeStr === 'string' && (array = timeStr.split(':')).length === 2) {
+        hour = parseInt(array[0])
+        minute = parseInt(array[1])
+      }
+    } catch (err) {
+      logger.warn(`dnd mode time paser error: ${timeStr}`)
+      hour = defaultHour
+      minute = defaultMinute
+    }
+    d.setHours(hour)
+    d.setMinutes(minute)
+    return d
+  }
+
+  /**
+   * Get the device active status
+   * @function isActivity
+   * @returns {string|null} app id if active or null if sleep
+   */
+  isActivity () {
+    return this.life.getCurrentAppId()
+  }
+}
+
+class DNDMode {
+  /**
+   * constructor of DNDMode
+   * @function constructor
+   * @param {object} runtime - app runtime
+   */
+  constructor (runtime) {
+    this.common = new DNDCommon(runtime.component.lifetime)
+    this.fsmStatus = FSM_READY
+    this.fsmTimer = undefined
+    this.waitSleep = false
+    this.fsmWaitingBreaker = undefined
+    this.isOptionBreaker = false
+    runtime.component.lifetime.on('idle', this.onDeviceIdle.bind(this))
+  }
+  /**
+   * when the device changes from active to idle, this function will be triggered
+   * @function onDeviceIdle
+   */
+  onDeviceIdle () {
+    if (this.waitSleep && this.fsmStatus === FSM_WAITING) {
+      this.waitSleep = false
+      if (this.fsmWaitingBreaker) {
+        logger.info('device is idle, recheck dnd mode')
+        this.fsmWaitingBreaker()
+      }
+    }
+  }
+  /**
+   * Get option from cloud
+   * @param {object} option dnd mode option
+   */
+  setOption (option) {
+    if (option === undefined) {
+      return
+    }
+    if (option.action !== undefined && typeof option.action === 'string' &&
```

Comment:
Redundant type check. `option.action === 'open'` should have covered both cases of `option.action !== undefined && typeof option.action === 'string'`.